Display Basic Course Information on Course Home Page

bradenmacdonald opened the pull request:
**Description**: This adds a new optional UI component to the "Home" page of each course in the Insights Dashboard, in order to display useful information about the course. This supersedes an earlier proposal, #331.

It displays:
* Course Name or ID (whichever is not shown already in the header)
* Course Start Date
* Course End Date
* Course Status (in progress, started, or ended)
* Links to the LMS and Studio (when configured)

**JIRA**: [OSPR-721](https://openedx.atlassian.net/browse/OSPR-721)

**Partner Info**: 3rd party Open edX instance. Submitting to minimize code drift but no merge requirement.

**Merge deadline**: Not an edX partner, so best effort.

**Screenshots**:

1. enable_course_overview on, display_course_name_in_nav on, course that hasn't yet started:
  ![not started yet](https://cloud.githubusercontent.com/assets/945577/8868589/22edd00e-318f-11e5-8367-0a1b2e8f1c93.png)

2. enable_course_overview on, display_course_name_in_nav off, course with no end date:
  ![in progress display_course_name_in_nav off](https://cloud.githubusercontent.com/assets/945577/8868590/2cb361a8-318f-11e5-839f-c153ce3672e5.png)
3. With enable_course_overview off, the page appears exactly the same as before.
![enable_course_overview off](https://cloud.githubusercontent.com/assets/945577/8868674/d1919b18-318f-11e5-89f7-974dbe2f10fb.png)


**Configuration**:
The new component is only shown if the waffle switch `enable_course_overview` is active. The information on the left side also requires `enable_course_api` to be active. The links on the right require two django settings, `LMS_COURSE_SHORTCUT_BASE_URL` (existing) and `CMS_COURSE_SHORTCUT_BASE_URL` (new).

**Test Instructions**:
1. Get a running copy of the dashboard.
2. In `settings/local.py`, set the values of `LMS_COURSE_SHORTCUT_BASE_URL` and `CMS_COURSE_SHORTCUT_BASE_URL`. If testing with devstack (e.g. if using the analytics devstack), set:

```python
LMS_COURSE_SHORTCUT_BASE_URL = 'http://localhost:8000/courses'
CMS_COURSE_SHORTCUT_BASE_URL = 'http://localhost:8001/course'
```

3. Go to {dashboard url}/admin/waffle/switch/ and activate various combinations of `enable_course_overview`, `display_course_name_in_nav`, and `enable_course_api`. Observe the effect on the home page of any course in the dashboard.

**Reviewers**: @mtyaka and TBD

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-analytics-dashboard&number=334

mtyaka commented on analytics_dashboard/courses/templates/courses/home.html:
```diff
@@ -29,6 +29,38 @@
+          <table class="table">
+            <tbody>
+              {% for label, value in course_overview %}
+                <tr><td>{{label}}</td><td>{{value|safe}}</td></tr>
```
Nit: semantically it would be nicer to use `<th>` instead of `<td>` for labels.

mtyaka commented on analytics_dashboard/courses/views/__init__.py:
```diff
@@ -524,6 +525,59 @@ def get_context_data(self, **kwargs):
 
         context['page_data'] = self.get_page_data(context)
+
+        external_tools = []
```
Can we move this closer to the code where it's first used?

mtyaka commented on analytics_dashboard/courses/views/__init__.py:
```diff
@@ -524,6 +525,59 @@ def get_context_data(self, **kwargs):
+                    return datetime.strptime(date_str, CourseStructureApiClient.DATETIME_FORMAT) if date_str else None
+
+                def format_date(date):
+                    return dateformat.format(date, settings.DATE_FORMAT) if date else "--"
```
Nit: I would put a newline here.

bradenmacdonald replied:
Sure, done.

mtyaka commented:
@bradenmacdonald I posted some nitpicks inline, but the changes look really good! :+1: Before merging this it would be good to document the new `enable_course_overview` switch the README. The existing switches are documented at: https://github.com/edx/edx-analytics-dashboard/blob/de57ca2667af4b888dd60d2986a52edebf576243/README.md#feature-gating

bradenmacdonald commented:
@mtyaka Ah, I didn't know where the switches were documented nor think to look. Thanks for pointing that out!

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! I've created [OSPR-721](https://openedx.atlassian.net/browse/OSPR-721) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

We can't start reviewing your pull request until you've added yourself to the [AUTHORS](https://github.com/edx/edx-analytics-dashboard/blob/master/AUTHORS) file. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst) file for more information.

shnayder commented:
Hi @bradenmacdonald. I like this feature a lot--it's definitely helpful to have the links back to the rest of the platform and the dates. I chatted with Andy on the UX team, and he suggested that since this info is less frequently necessary than the links for diving into the data, we should move the overview section below the navigational section. Thoughts on that?

A few small UI nits:
- use the same header styles as elsewhere on the page ("Course Overview" and "External tools" should look the same as "Enrollment", "Engagement", and "Performance")
- we suggest getting rid of the lines between rows in the overview section
- Make sure the links to Studio + LMS are styled the same as the other links on the page 

Finally, I would want to just have this always present, and not have a separate feature flag for it unless it's necessary.

bradenmacdonald commented:
@shnayder Yep, all of your comments make sense. I suspected I would be asked to move this below the other links - I probably should have just gone with that in the first place, but I was torn.

I've pushed a commit to address your feedback. I removed the new waffle switch, though the feature will still disappear if the course structure API is turned off.

I also made the UX tweaks. I wasn't quite clear on the headers - so I went with the same headers seen on the sub-pages like activity, demographics, etc. If you meant bold grey headers or even to put the info in white boxes with grey headers let me know.

Updated s/s:
![screen shot 2015-07-28 at 10 44 22 pm](https://cloud.githubusercontent.com/assets/945577/8950741/5dce275c-357a-11e5-9334-28837918147b.png)

shnayder commented:
Thanks for the quick response. I like the styling--I was originally thinking of the gray headers, but that wouldn't work without the white background.

I'll get this on the list for the team to review.

dsjen commented on analytics_dashboard/courses/templates/courses/home.html:
```diff
@@ -60,4 +60,35 @@
+    </div>
+
+    <div class="row">
+      <div class="col-sm-4">
```
What do you think about using `col-sm-8` rather and removing the `col-sm-offset-4` css class in the section below?  That would allow the table to not wrap in smaller screens.

dsjen commented on analytics_dashboard/courses/templates/courses/home.html:
```diff
@@ -60,4 +60,35 @@
+        {% endif %}
+      </div>
+      {% if external_course_tools %}
+        <div class="col-sm-4 col-sm-offset-4">
```
A translator note would be helpful here (something like, `{#  Translators: This title lists links to outside tools for this course. #}`)

dsjen commented on analytics_dashboard/courses/templates/courses/home.html:
```diff
@@ -60,4 +60,35 @@
+      <div class="col-sm-4">
+        {% if course_overview %}
+          <h4 class="section-title">{% trans "Course Overview" %}</h4>
+          <table class="table course-overview">
```
Is the `table` class necessary?  If needed as a CSS selector, it seems as though that could happen through the DOM element itself.

dsjen replied:
The data represented isn't tabular, so I'd recommend not using a table (also raises accessibility concerns for page navigation).  Rather, some nested columns could be used to accomplish the look--something like:

```
{% for label, value in course_overview %}
  <div class="row">
    <div class="col-xs-4">{{ label }}</div>
    <div class="col-xs-8">{{ value|safe }}</div>
  </div>
```

bradenmacdonald replied:
`.table` is a Bootstrap class used to bring in a bunch of nice styling. It's necessary to distinguish tables that present information from tables [ab]used for document layout/structure.

bradenmacdonald replied:
@dsjen Taking a look at this now... Here's how it looks when I convert to columns with the HTML you suggest above, instead of a table (though I have not adjusted the spacing):
![screen shot 2015-11-03 at 12 22 42 pm](https://cloud.githubusercontent.com/assets/945577/10920578/ac67e810-8225-11e5-9c08-ed0d6f17e18c.png)

However, when it's not a table, it's missing semantic information. To a screen reader, wouldn't this collection of random DIVs just appear as "Name edX Demonstration Course Start Date February 05, 2013 ..." ?

The other issue is that a table will automatically adjust the column widths based on the content. Whereas with the column approach, I need to set a column width based on what looks good in English (aiming to keep the left column as narrow as possible without word wrapping, for ease of reading), which won't adapt well when the text is translated to a language where these phrases are much longer.

dsjen replied:
Thanks for the screenshot!

@cptvitamin, do you know what the best way to represent this information is?  I recommended not using a table, but am not entirely sure now.

cptvitamin replied:
@bradenmacdonald @dsjen Great question!  So, html ```<table>```s got a really bad rap back in the day when designers started using them to do page layout.  They were designed only to be used for tables that displayed tabular data, like a spreadsheet.  CSS could solve the problem using floats, but that caused other problems, especially for screen reader reading order when the DOM order was different than the visual display order.  flex-box and CSS grids came on the scene to solve that problem but they are quite complicated.  Long story short, as long as you mark your table with a role=presentation like this ```<table role="presentation">``` its fine to use tables to do simple layouts like this.  the role definition causes the screen reader to treat each of these as block level elements, presented in DOM order.  Either way, tables or divs, WFM.

dsjen replied:
Thank you!  That was enlightening.  @bradenmacdonald -- looks like all we need is to add in `role="presentation"`.

dpowell replied:
Consider using a `dl` element.

https://html.spec.whatwg.org/multipage/semantics.html#the-dl-element
http://getbootstrap.com/css/#horizontal-description

bradenmacdonald replied:
@dpowell That's a great suggestion, thanks. I always forget about the `dl` element. In this case, I actually chose "table" not primarily for layout purposes but because I thought it represented the semantics better than plain divs, with `th` headers and corresponding `td` data for each header. (And a two-column table like a traditional "Table of Contents" is still arguably a table.) But I think `dl` is the best fit for this type of information.

So: I tried `dl` and it looks like this:
![screen shot 2015-11-03 at 1 49 35 pm](https://cloud.githubusercontent.com/assets/945577/10922809/30879ddc-8232-11e5-90d8-e8ab514a0bc6.png)

or this:
![screen shot 2015-11-03 at 1 49 20 pm](https://cloud.githubusercontent.com/assets/945577/10922814/3686508e-8232-11e5-8e08-2283b35e364b.png)

I'm not a huge fan of the look of either of those, but I could override it with a bunch of CSS to have the same display as before.

bradenmacdonald replied:
Ok after looking a bit more, `dl` ends up very similar to table but we lose the ability to have the browser resize the columns based on their content (`dt` requires fixed widths for a column-like appearance). And as [this guy](http://stackoverflow.com/a/9621230/1057326) argues, the semantics of `dl` and the sort of `th-td` table I had are basically identical.

dsjen commented on analytics_dashboard/static/sass/_developer.scss:
```diff
@@ -395,6 +395,27 @@ table.dataTable thead th.sorting_desc:after {
   }
 }
 
+.table.course-overview {
```
Instead of a class, `table` itself could be the selector.

dsjen commented on analytics_dashboard/courses/templates/courses/home.html:
```diff
@@ -60,4 +60,35 @@
+    <div class="row">
+      <div class="col-sm-4">
+        {% if course_overview %}
+          <h4 class="section-title">{% trans "Course Overview" %}</h4>
```
I chatted with @lamagnifica and she suggested the title reflect the course information section in the instructor dashboard (see attached screenshot for example) and the title be updated to "Basic Course Information" instead.

![screen shot 2015-10-23 at 2 53 12 pm](https://cloud.githubusercontent.com/assets/5265058/10702142/d41f26f4-7995-11e5-8315-966b42cd37e8.png)

bradenmacdonald replied:
Sure thing.

dsjen commented on analytics_dashboard/courses/views/__init__.py:
```diff
@@ -524,6 +525,59 @@ def get_context_data(self, **kwargs):
+        overview_data = []
+        if self.course_api_enabled:
+            if switch_is_active('display_course_name_in_nav'):
+                overview_data.append((_('ID'), self.course_id))
```
Can this be updated to "Course ID" instead?  A translator note with a note that ID stands for "identifier" would be helpful too.

dsjen commented on analytics_dashboard/courses/views/__init__.py:
```diff
@@ -524,6 +525,59 @@ def get_context_data(self, **kwargs):
+            if switch_is_active('display_course_name_in_nav'):
+                overview_data.append((_('ID'), self.course_id))
+            else:
+                overview_data.append((_('Name'), self.course_info.get('name')))
```
Can this be updated to "Course Name" instead?

dsjen commented on analytics_dashboard/courses/views/__init__.py:
```diff
@@ -524,6 +525,59 @@ def get_context_data(self, **kwargs):
+            if start_date:
+                if todays_date >= start_date:
+                    in_progress = (end_date is None or end_date > todays_date)
+                    status_str = _('In Progress') if in_progress else _('Ended')
```
Translator notes that these are course statuses would be helpful.

dsjen commented on analytics_dashboard/courses/views/__init__.py:
```diff
@@ -524,6 +525,59 @@ def get_context_data(self, **kwargs):
+                    in_progress = (end_date is None or end_date > todays_date)
+                    status_str = _('In Progress') if in_progress else _('Ended')
+                else:
+                    status_str = _('Not Started Yet')
```
Translator notes would be helpful.

dsjen commented on analytics_dashboard/courses/views/__init__.py:
```diff
@@ -524,6 +525,59 @@ def get_context_data(self, **kwargs):
+            })
+        if settings.CMS_COURSE_SHORTCUT_BASE_URL:
+            external_tools.append({
+                'title': _('Studio'),
```
@shnayder -- is Studio called "Studio" in other edX instances?

shnayder replied:
Yes, I believe so. Braden would know better than me, probably :)

bradenmacdonald replied:
The name ("Studio") can be customized on each deployment, but AFAIK there is no way for the Insights dashboard to determine what the user's particular installation calls their CMS. "Studio" is the default name.

dsjen replied:
Ok, thanks!  I'm wondering if we should translate it or perhaps add in a translator comment--no strong preference here.

dsjen commented:
@bradenmacdonald -- thank you for this PR!  I tested it out and it works as promised.  I left a new notes in the review.  Please let me know if you have questions.

dsjen commented:
@bradenmacdonald -- I just wanted to check in about this PR.  Is there a plan to merge it in?  Do you need anything from me?  Thanks!

bradenmacdonald commented:
@dsjen Thanks for the reminder! I'll rebase and make those minor tweaks on Monday.

bradenmacdonald commented:
@dsjen Thanks for the review (and sorry for the delay in addressing your comments!). I've pushed 34f9b6489f368305fa68ce8118b8d54b5a2cb42c to address everything you raised. The only thing I didn't change is the `table` conversion to columns. See my response to your comments on that and let me know your preference please :)

dsjen commented on analytics_dashboard/courses/views/__init__.py:
```diff
@@ -525,6 +526,62 @@ def get_context_data(self, **kwargs):
+                if todays_date >= start_date:
+                    in_progress = (end_date is None or end_date > todays_date)
+                    # Translators: These refer to whether students are actively using the course or it is over.
+                    status_str = _('In Progress') if in_progress else _('Ended')
```
@sarina, will this translator comment apply to both strings?  Would it be better for them to be separated into variables?

sarina replied:
This comment will be attached to both strings, without reference to either. It's fine as is but I'd replace the word "these" with "'In Progress' and 'Ended' refer to..." for extra clarity.

dsjen replied:
Ok, thank you!

dsjen commented on analytics_dashboard/courses/views/__init__.py:
```diff
@@ -525,6 +526,62 @@ def get_context_data(self, **kwargs):
+            })
+        if settings.CMS_COURSE_SHORTCUT_BASE_URL:
+            external_tools.append({
+                'title': _('Studio'),  # Translators: This is the name of the edX Course Editor (the CMS)
```
@sarina, is it recommended that translator comments go above the string?

bradenmacdonald commented:
@dsjen Ok I just made this change (f1cef421f1db8137e3f80df77b76755582ce2ad8) and then squashed the commits. Going with the `table role=presentation` for now, so this is ready to merge if you like that and the tests pass.

dsjen commented:
@bradenmacdonald -- thank you!  Please merge when tests pass.  :+1:

sarina commented on analytics_dashboard/courses/views/__init__.py:
```diff
@@ -525,6 +526,64 @@ def get_context_data(self, **kwargs):
+        if settings.CMS_COURSE_SHORTCUT_BASE_URL:
+            external_tools.append({
+                # Translators: This is the name of the edX Course Editor (the CMS)
+                'title': _('Studio'),
```
We have a variable, settings.STUDIO_NAME I think??? you should use that. Studio is a brand name that is not translated.

sarina replied:
yes, there's both `settings.STUDIO_NAME` and `settings.STUDIO_SHORT_NAME` (I believe "edX Studio" and "Studio", respectively, on our servers).

sarina replied:
@bradenmacdonald I know you just merged this, but this isn't a string we should be serving to translators.

bradenmacdonald replied:
@sarina I don't mind doing a fixup PR, but as far as I know, there is no way to access that setting from the analytics dashboard (Insights) django app. It's not in the Insights settings files nor in the ansible-generated `/edx/etc/insights.yml` environment.

bradenmacdonald replied:
@sarina I could just mark this as not translated though.
